### PR TITLE
lint: Convert lint-include-guards.sh to Python

### PR DESCRIPTION
**Pull request by brydinh**
This PR addresses [issue 24783](https://github.com/bitcoin/bitcoin/issues/24783).  Converted lint-include-guards.sh to python.

### CODE REVIEW

**Comment by theStack** [comment]
Concept ACK

Note that there is currently a linter error in the CI:
```
Traceback (most recent call last):
  File "test/lint/lint-include-guards.py", line 88, in <module>
    main()
  File "test/lint/lint-include-guards.py", line 71, in main
    header_file_contents_string = open(header_file, 'r').read()
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xce in position 2003: ordinal not in range(128)
^---- failure generated from test/lint/lint-include-guards.py
Python's open(...) seems to be used to open text files without explicitly
specifying encoding="utf8":

test/lint/lint-include-guards.py:        header_file_contents_string = open(header_file, 'r').read()
^---- failure generated from test/lint/lint-python-utf8-encoding.sh
Success: no issues found in 236 source files
```

File: test/lint/lint-include-guards.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+def main():
+    exit_code = 0
+
+    header_file_lst = _get_header_file_lst()
+    for header_file in header_file_lst:
+        header_id = _get_header_id(header_file)
+
+        regex_pattern = f'#(ifndef|define|endif //) {header_id}'
+
+        header_file_contents_string = open(header_file, 'r').read()
```
Comment (brunoerg) [inline test/lint/lint-include-guards.py:71]: I prefer to open files using `with`, it ensures that the file will be closed at the end of the operation.

Comment (brydinh) [reply test/lint/lint-include-guards.py:71]: Good catch!

**Comment by KevinMusgrave** [comment]
I tested ea3e7abc45730e7a6eb1e840c6fa40780b3dee2f by commenting out the include guard in `bench/bench.h` and got the following results:

Before (`test/lint/lint-include-guards.sh`):
```
src/bench/bench.h seems to be missing the expected include guard:
  #ifndef BITCOIN_BENCH_BENCH_H
  #define BITCOIN_BENCH_BENCH_H
  ...
  #endif // BITCOIN_BENCH_BENCH_H

```

After (`test/lint/lint-include-guards.py`):

The test passes (nothing prints)

File: test/lint/lint-include-guards.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        regex_pattern = f'#(ifndef|define|endif //) {header_id}'
+
+        with open(header_file, 'r', encoding='utf-8') as f:
+            header_file_contents_string = f.read()
+
+        include_guard_lst = re.findall(
+            regex_pattern, header_file_contents_string)
+
+        if len(include_guard_lst) != 3:
```
Comment (KevinMusgrave) [inline test/lint/lint-include-guards.py:83]: The following is one way to fix it:

```suggestion
        regex_pattern = f'^#(ifndef|define|endif //) {header_id}'

        with open(header_file, 'r', encoding='utf-8') as f:
            header_file_contents = f.readlines()

        count = 0
        for header_file_contents_string in header_file_contents:
            include_guard_lst = re.findall(
                regex_pattern, header_file_contents_string)
            
            count += len(include_guard_lst)

        if count != 3:
```

Comment (brydinh) [reply test/lint/lint-include-guards.py:83]: thanks, awesome catch!

**Comment by KevinMusgrave** [comment]
Tested ACK d5fdec5cf8cffe8ece5460cd8c6e83ea6eebf625

Outputs and exit codes match. Example:
```
src/attributes.h seems to be missing the expected include guard:
  #ifndef BITCOIN_ATTRIBUTES_H
  #define BITCOIN_ATTRIBUTES_H
  ...
  #endif // BITCOIN_ATTRIBUTES_H

src/banman.h seems to be missing the expected include guard:
  #ifndef BITCOIN_BANMAN_H
  #define BITCOIN_BANMAN_H
  ...
  #endif // BITCOIN_BANMAN_H

src/bech32.h seems to be missing the expected include guard:
  #ifndef BITCOIN_BECH32_H
  #define BITCOIN_BECH32_H
  ...
  #endif // BITCOIN_BECH32_H
```

**Comment by laanwj** [comment]
Code review ACK d5fdec5cf8cffe8ece5460cd8c6e83ea6eebf625

**Pull request merged**